README.md
# Croninfo

[![PyPi Version](https://img.shields.io/pypi/v/croninfo.svg?style=flat-square&logo=PyPi)](https://pypi.org/project/croninfo/)
[![PyPi License](https://img.shields.io/pypi/l/croninfo.svg?style=flat-square)](https://pypi.org/project/croninfo/)
[![CI
Tests](https://github.com/paulmonk/croninfo/workflows/CI%20Test/badge.svg)](https://github.com/paulmonk/croninfo/actions?query=workflow%3A%22CI+Test%22)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square)](https://github.com/psf/black)

**Croninfo** is a CLI which provides the functionality of parsing
cron expressions and rendering an output to ease with understanding the schedule
of the cron expression.

The Cron syntax supported is based on the [definition provided by FreeBSD](https://www.freebsd.org/cgi/man.cgi?crontab%285%29).

# Getting Started

**Python 3.7 to 3.10 supported**

To get started, install with pip:

```shell
$ python -m pip install croninfo
```

# Usage

## CLI

To CLI is provided with one command `parse` as can be seen below.
If any doubts you can run `croninfo --help` or `croninfo <command> --help`
for further details.

```shell
$ croninfo parse "10 0 1,15 * 1-3 /usr/bin/find"
╭─ Cron Expression ──────────────────────────────────────────────────────────────────────────────╮
│ Minute               10                                                                        │
│ Hour                 0                                                                         │
│ Day of Month         1 15                                                                      │
│ Month                1 2 3 4 5 6 7 8 9 10 11 12                                                │
│ Day of Week          1 2 3                                                                     │
│ TZ                   UTC                                                                       │
│ Command              /usr/bin/find                                                             │
│ Next Scheduled Run   2022-08-15T00:10:00+00:00 (in 10 days, 2 hours, 3 minutes and 25 seconds) │
╰─ 10 0 1,15 * 1-3 /usr/bin/find ────────────────────────────────────────────────────────────────╯
```

This will output with consideration to **UTC** by default but you can pass
`--tz-type local` to override to be your local timezone, E.G. `Europe/London`.
As shown below:

```shell
$ croninfo parse "10 0 1,15 * 1-3 /usr/bin/find" --tz-type local
╭─ Cron Expression ──────────────────────────────────────────────────────────────────────────────╮
│ Minute               10                                                                        │
│ Hour                 0                                                                         │
│ Day of Month         1 15                                                                      │
│ Month                1 2 3 4 5 6 7 8 9 10 11 12                                                │
│ Day of Week          1 2 3                                                                     │
│ TZ                   Europe/London                                                             │
│ Command              /usr/bin/find                                                             │
│ Next Scheduled Run   2022-08-15T00:10:00+01:00 (in 10 days, 1 hour, 2 minutes and 25 seconds)  │
╰─ 10 0 1,15 * 1-3 /usr/bin/find ────────────────────────────────────────────────────────────────╯
```

Alternatively, you can use the [**Official Croninfo Docker Image**](https://hub.docker.com/r/paulmonk/croninfo)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2022-08-05

### Fixed

- PyPi Build

## [1.0.0] - 2022-08-05 [YANKED]

### Added

- Initial Commit!


tests/conftest.py
from __future__ import annotations

from functools import partial
from textwrap import dedent

import pytest
from typer.testing import CliRunner


def assert_cli_output(result, output):
    assert result.output.strip() == dedent(output).strip()


@pytest.fixture(scope="session")
def typer_runner():
    runner = CliRunner()

    def inner(cmd, *args, **kwargs):
        result = runner.invoke(cmd, *args, **kwargs)
        result.assert_cli_output = partial(assert_cli_output, result)
        return result

    return inner


tests/test_cli.py
from __future__ import annotations

import datetime as dt
import sys

import pytest
import time_machine

from croninfo.cli import __version__, cli

# Backports is required for Python versions <3.9
if sys.version_info >= (3, 9):
    import zoneinfo
else:
    from backports import zoneinfo


@pytest.mark.parametrize(
    "expr, expected",
    [
        pytest.param(
            ["@every_minute /usr/bin/find"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 │
            │ 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46   │
            │ 47 48 49 50 51 52 53 54 55 56 57 58 59                                       │
            │ Hour                 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 │
            │ 22 23                                                                        │
            │ Day of Month         1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21   │
            │ 22 23 24 25 26 27 28 29 30 31                                                │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          1 2 3 4 5 6 7                                           │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-01-01T01:01:00+00:00 (in less than 60 seconds)     │
            ╰─ @every_minute /usr/bin/find ────────────────────────────────────────────────╯
            """,
            id="Less than 60 Seconds",
        ),
        pytest.param(
            ["@weekly /usr/bin/find"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0                                                       │
            │ Hour                 0                                                       │
            │ Day of Month         1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21   │
            │ 22 23 24 25 26 27 28 29 30 31                                                │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          7                                                       │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-01-02T00:00:00+00:00 (in 22 hours, 58 minutes and  │
            │ 58 seconds)                                                                  │
            ╰─ @weekly /usr/bin/find ──────────────────────────────────────────────────────╯
            """,
            id="Hours, Minutes, Seconds",
        ),
        pytest.param(
            ["*/2 * * * * /usr/bin/find"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38  │
            │ 40 42 44 46 48 50 52 54 56 58                                                │
            │ Hour                 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 │
            │ 22 23                                                                        │
            │ Day of Month         1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21   │
            │ 22 23 24 25 26 27 28 29 30 31                                                │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          1 2 3 4 5 6 7                                           │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-01-01T01:02:00+00:00 (in 58 seconds)               │
            ╰─ */2 * * * * /usr/bin/find ──────────────────────────────────────────────────╯
            """,
            id="Seconds only",
        ),
        pytest.param(
            ["18-25 6-8,10-12 * JAN-DEC/2 SUN,TUE-fri /usr/bin/find"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               18 19 20 21 22 23 24 25                                 │
            │ Hour                 6 7 8 10 11 12                                          │
            │ Day of Month         1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21   │
            │ 22 23 24 25 26 27 28 29 30 31                                                │
            │ Month                1 3 5 7 9 11                                            │
            │ Day of Week          2 3 4 5 7                                               │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-01-02T06:18:00+00:00 (in 1 day, 5 hours, 16        │
            │ minutes and 58 seconds)                                                      │
            ╰─ 18-25 6-8,10-12 * JAN-DEC/2 SUN,TUE-fri /usr/bin/find ──────────────────────╯
            """,
            id="Day",
        ),
        pytest.param(
            ["18-25 1-2 1 JAN-DEC/2 SUN,TUE-fri /usr/bin/find"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               18 19 20 21 22 23 24 25                                 │
            │ Hour                 1 2                                                     │
            │ Day of Month         1                                                       │
            │ Month                1 3 5 7 9 11                                            │
            │ Day of Week          2 3 4 5 7                                               │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-03-01T01:18:00+00:00 (in 59 days, 16 minutes and   │
            │ 58 seconds)                                                                  │
            ╰─ 18-25 1-2 1 JAN-DEC/2 SUN,TUE-fri /usr/bin/find ────────────────────────────╯
            """,
            id="Days and No Hour",
        ),
    ],
)
@time_machine.travel(
    dt.datetime(
        year=2022, month=1, day=1, hour=1, minute=1, second=1, tzinfo=dt.timezone.utc
    )
)
def test_parse_command__output(expr, expected, typer_runner):
    """
    Given any type of valid cron expression and tz args input expect the output
    to be correct and formatted as desired.
    """
    cmds = ["parse"]
    cmds.extend(expr)
    result = typer_runner(cli, cmds)

    assert 0 == result.exit_code
    result.assert_cli_output(expected)


@pytest.mark.parametrize(
    "expr, expected, tzinfo",
    [
        pytest.param(
            ["*/15 0 1,15 * 1-5 /usr/bin/find"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0 15 30 45                                              │
            │ Hour                 0                                                       │
            │ Day of Month         1 15                                                    │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          1 2 3 4 5                                               │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-02-01T00:00:00+00:00 (in 30 days, 22 hours, 58     │
            │ minutes and 58 seconds)                                                      │
            ╰─ */15 0 1,15 * 1-5 /usr/bin/find ────────────────────────────────────────────╯
            """,
            zoneinfo.ZoneInfo("Europe/London"),
            id="Test UTC Opt Default",
        ),
        pytest.param(
            ["*/15 0 1,15 * 1-5 /usr/bin/find", "--tz-type", "utc"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0 15 30 45                                              │
            │ Hour                 0                                                       │
            │ Day of Month         1 15                                                    │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          1 2 3 4 5                                               │
            │ TZ                   UTC                                                     │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-02-01T00:00:00+00:00 (in 30 days, 22 hours, 58     │
            │ minutes and 58 seconds)                                                      │
            ╰─ */15 0 1,15 * 1-5 /usr/bin/find ────────────────────────────────────────────╯
            """,
            zoneinfo.ZoneInfo("Europe/London"),
            id="Test UTC Opt",
        ),
        pytest.param(
            ["*/15 0 1,15 * 1-5 /usr/bin/find", "--tz-type", "local"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0 15 30 45                                              │
            │ Hour                 0                                                       │
            │ Day of Month         1 15                                                    │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          1 2 3 4 5                                               │
            │ TZ                   Europe/London                                           │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-02-01T00:00:00+00:00 (in 30 days, 22 hours, 58     │
            │ minutes and 58 seconds)                                                      │
            ╰─ */15 0 1,15 * 1-5 /usr/bin/find ────────────────────────────────────────────╯
            """,
            zoneinfo.ZoneInfo("Europe/London"),
            id="Test Local Opt - London",
        ),
        pytest.param(
            ["*/15 0 1,15 * 1-5 /usr/bin/find", "--tz-type", "local"],
            """
            ╭─ Cron Expression ────────────────────────────────────────────────────────────╮
            │ Minute               0 15 30 45                                              │
            │ Hour                 0                                                       │
            │ Day of Month         1 15                                                    │
            │ Month                1 2 3 4 5 6 7 8 9 10 11 12                              │
            │ Day of Week          1 2 3 4 5                                               │
            │ TZ                   Europe/Berlin                                           │
            │ Command              /usr/bin/find                                           │
            │ Next Scheduled Run   2022-02-01T00:00:00+01:00 (in 30 days, 21 hours, 58     │
            │ minutes and 58 seconds)                                                      │
            ╰─ */15 0 1,15 * 1-5 /usr/bin/find ────────────────────────────────────────────╯
            """,
            zoneinfo.ZoneInfo("Europe/Berlin"),
            id="Test Local Opt - Berlin",
        ),
    ],
)
@time_machine.travel(
    dt.datetime(
        year=2022,
        month=1,
        day=1,
        hour=1,
        minute=1,
        second=1,
        tzinfo=dt.timezone.utc,
    )
)
def test_parse_command__opts_output(expr, expected, tzinfo, typer_runner, mocker):
    """
    Given any type of valid cron expression and tz args input expect the output
    to be correct and formatted as desired.
    """
    cmds = ["parse"]
    cmds.extend(expr)

    # Ensure tzlocal always picks up the expected tzinfo being passed.
    mocker.patch("croninfo.cli.tzlocal.get_localzone", return_value=tzinfo)

    result = typer_runner(cli, cmds)

    assert 0 == result.exit_code
    result.assert_cli_output(expected)


@pytest.mark.parametrize(
    "expr",
    [
        (["blah"]),
        (["*/15 0 1,15 * /usr/bin/find"]),
    ],
)
def test_parse_command__invalid_expressions(expr, typer_runner):
    """
    Given any type of invalid cron expression input expect an error to be
    returned.
    """
    cmds = ["parse"]
    cmds.extend(expr)
    result = typer_runner(cli, cmds)

    assert 1 == result.exit_code


@pytest.mark.parametrize(
    "expr, expected",
    [
        (
            ["*/15 0 1,15 * 1-5 /usr/bin/find", "--tz-type", "London"],
            "Invalid value for '--tz-type': 'London' is not one of 'local', 'utc'",
        ),
        (
            ["*/15 0 1,15 * 1-12 /usr/bin/test", "--tz-type", "Berlin"],
            "Invalid value for '--tz-type': 'Berlin' is not one of 'local', 'utc'",
        ),
    ],
)
def test_parse_command__invalid_opts(expr, expected, typer_runner):
    """
    Given any type of invalid cron expression input expect an error to be
    returned.
    """
    cmds = ["parse"]
    cmds.extend(expr)
    result = typer_runner(cli, cmds)

    assert 2 == result.exit_code
    assert expected in str(result.stdout)


def test_version_arg(typer_runner):
    """
    The --version output returns the expected version.
    """
    result = typer_runner(cli, "--version")

    assert 0 == result.exit_code
    result.assert_cli_output(f"Version: {__version__}")


def test_version_arg__is_eager(typer_runner):
    """
    The --version arg should be eager and supersede any commands if supplied.
    """
    result = typer_runner(cli, ["--version", "parse", "blah"])

    assert 0 == result.exit_code
    result.assert_cli_output(f"Version: {__version__}")


src/croninfo/cli.py
from __future__ import annotations

import datetime as dt
from enum import Enum

import typer
import tzlocal
from rich.console import Console
from rich.panel import Panel

from croninfo import __version__
from croninfo.crontab import Crontab

cli = typer.Typer()


class ParseTZOpts(str, Enum):
    LOCAL = "local"
    UTC = "utc"


@cli.command()
def parse(
    expression: str,
    tz_type: ParseTZOpts = typer.Option(  # noqa: B008
        ParseTZOpts.UTC.value, "--tz-type", case_sensitive=False
    ),
) -> None:
    """
    Accept the input of a Crontab expression, which is then parsed into a data structure.
    All datetime info is parsed in the timezone provided, defaults to UTC.
    """
    tz = (
        dt.timezone.utc
        if tz_type.value == ParseTZOpts.UTC.value
        else tzlocal.get_localzone()
    )
    crontab = Crontab.from_parse(expr=expression, tz=tz)

    # Determine next scheduled run of crontab.
    next_run = crontab.next_scheduled_run
    next_run_delta = next_run - dt.datetime.now(tz=crontab.tz)

    # Map of desired header name->value
    output_map = {
        "Minute": crontab.minute.values,
        "Hour": crontab.hour.values,
        "Day of Month": crontab.monthday.values,
        "Month": crontab.month.values,
        "Day of Week": crontab.weekday.values,
        "TZ": crontab.tz,
        "Command": crontab.command,
        "Next Scheduled Run": f"{next_run.isoformat()} (in {_format_friendly_timedelta(next_run_delta)})",
    }
    final_output = []
    for k, v in output_map.items():
        # 'Command' value is a string and we do not want the join iteration to apply to
        # strings which should still be outputted as the original value.
        result = v
        if isinstance(v, list):
            result = " ".join(str(x) for x in v if v)
        final_output.append(f"[bold]{k:<20}[/bold] {result}")

    panel = Panel(
        "\n".join(final_output),
        title="Cron Expression",
        title_align="left",
        subtitle=expression,
        subtitle_align="left",
        expand=False,
    )
    console = Console()
    console.print(panel, justify="left")


def _format_friendly_timedelta(delta: dt.timedelta) -> str:
    days = delta.days

    # In cases where delta days can be -1 short-circuit.
    # E.G. cron is scheduled to run every minute.
    if days < 0:
        return "less than 60 seconds"

    hours, _rem = divmod(delta.seconds, 3600)
    mins, secs = divmod(_rem, 60)

    # Avoid returns extra noise when specific fragments are not required.
    descriptor = "seconds" if secs > 1 else "second"
    pattern = f"{secs:.0f} {descriptor}"
    if mins > 0:
        descriptor = "minutes" if mins > 1 else "minute"
        pattern = f"{mins:.0f} {descriptor} and {pattern}"
    if hours > 0:
        descriptor = "hours" if hours > 1 else "hour"
        pattern = f"{hours:.0f} {descriptor}, {pattern}"
    if days > 0:
        descriptor = "days" if days > 1 else "day"
        pattern = f"{days} {descriptor}, {pattern}"
    return pattern


def _version_callback(value: bool) -> None:
    if value:
        typer.echo(f"Version: {__version__}")
        raise typer.Exit()


@cli.callback()
def main(
    ctx: typer.Context,
    version: bool = typer.Option(  # noqa: B008
        None, "--version", callback=_version_callback, is_eager=True
    ),
) -> None:
    """
    A CLI tool which accepts a Crontab expression to be parsed with output.
    See "parse" command for more information.
    """


src/croninfo/crontab.py
from __future__ import annotations

import calendar
import dataclasses
import datetime as dt
from typing import ClassVar, Iterator

# Definitions based on spec here:
# https://www.freebsd.org/cgi/man.cgi?crontab%285%29
# @reboot and @every_second have been omitted.
CRON_MACROS = {
    "@yearly": "0 0 1 1 *",
    "@annually": "0 0 1 1 *",
    "@monthly": "0 0 1 * *",
    "@weekly": "0 0 * * 0",
    "@daily": "0 0 * * *",
    "@midnight": "0 0 * * *",
    "@hourly": "0 * * * *",
    "@every_minute": "*/1 * * * *",
}


@dataclasses.dataclass(frozen=True)
class CronPart:
    """
    Definition of a Crontab part.
    """

    # Friendly name for Part.
    name: ClassVar[str]

    # Each Cron part has a specific min/max range that is valid for it's part.
    min_value: ClassVar[int]
    max_value: ClassVar[int]

    # A cron part can have word aliases which convert to integers. E.G Months
    aliases: ClassVar[dict[str, int]]

    # Parsed value of the cron expression.
    values: list[int]

    @classmethod
    def __init_subclass__(
        cls,
        *,
        name: str,
        min_value: int,
        max_value: int,
        aliases: dict[str, int] | None = None,
    ):
        cls.name = name
        cls.min_value = min_value
        cls.max_value = max_value
        cls.aliases = aliases or {}

    def __iter__(self) -> Iterator[int]:
        return iter(self.values)

    def __str__(self) -> str:
        return str(self.values)

    @classmethod
    def from_expr(cls, expr: str):  # type: ignore
        """
        Parses crontab expression part based on rules provided (min-value, max-value, aliases etc...)

        *   any value
        ,   value list separator
        -   range of values
        /   step values
        """
        values = cls._parse(expr=expr)
        return cls(values=values)

    @classmethod
    def _parse(cls, *, expr: str, step: int = 1) -> list[int]:
        # Expressions can be delimited by commas.
        if "," in expr:
            # "parse" needs to be recursed to capture all possible values.
            values = {
                x
                for sub_expr in expr.split(",", maxsplit=2)
                for x in cls._parse(expr=sub_expr, step=step)
            }
            return list(values)

        # Wildcard, need to return all possible values
        elif expr == "*":
            # +1 as range arg are 0-based and we need result to be 1-based.
            return list(range(cls.min_value, cls.max_value + 1, step))

        # Single numeric value
        elif expr.isnumeric():
            return [cls._try_parse_int(expr)]

        # Step values
        # The rhs of an expression containing a / is the "step" value,
        # i.e `*/8` means every 8 minutes
        elif "/" in expr:
            try:
                lhs, rhs = expr.split("/", maxsplit=2)
            except ValueError:
                raise ValueError(
                    f"{cls.name} value must not contain more than one step parameter (/)"
                )

            rhs_step = cls._try_parse_int(rhs)
            return cls._parse(expr=lhs, step=rhs_step)

        # Expressions here represent two values
        elif "-" in expr:
            try:
                range_start, range_end = expr.split("-", maxsplit=2)
            except ValueError:
                raise ValueError(
                    f"{cls.name} value must not contain more than one range parameter (-)"
                )

            # x-y. x must be less than y, valid this.
            start = cls._try_parse_int(range_start)
            end = cls._try_parse_int(range_end)
            if start > end:
                raise ValueError(
                    f"{cls.name} range start value must not be > than end value"
                )
            return list(
                # +1 as range arg are 0-based and we need result to be 1-based.
                range(
                    start,
                    end + 1,
                    step,
                )
            )
        # Parse any remaining values. Will catch aliases here.
        else:
            return [cls._try_parse_int(expr)]

    @classmethod
    def _try_parse_int(cls, value: str | int) -> int:
        x = value

        # Check if the value is an alias that needs to be mapped.
        if isinstance(value, str):
            try:
                # Case should not matter for matches.
                x = cls.aliases[value.upper()]
            except KeyError:
                pass

        # Check value is a valid integer.
        try:
            x = int(x)
        except ValueError:
            raise ValueError(f"{cls.name} value must be of type int")

        if not (cls.min_value <= x <= cls.max_value):
            raise ValueError(
                f"{cls.name} value must be in range of "
                f"[{cls.min_value}, {cls.max_value}]"
            )
        return x


@dataclasses.dataclass(frozen=True)
class CronPartMinute(CronPart, name="Minute", min_value=0, max_value=59):
    """
    Crontab expression minute part.
    """


@dataclasses.dataclass(frozen=True)
class CronPartHour(CronPart, name="Hour", min_value=0, max_value=23):
    """
    Crontab expression hour part.
    """


@dataclasses.dataclass(frozen=True)
class CronPartMonthday(CronPart, name="Monthday", min_value=1, max_value=31):
    """
    Crontab expression day part.
    """


@dataclasses.dataclass(frozen=True)
class CronPartMonth(
    CronPart,
    name="Month",
    min_value=1,
    max_value=12,
    aliases={
        "JAN": 1,
        "FEB": 2,
        "MAR": 3,
        "APR": 4,
        "MAY": 5,
        "JUN": 6,
        "JUL": 7,
        "AUG": 8,
        "SEP": 9,
        "OCT": 10,
        "NOV": 11,
        "DEC": 12,
    },
):
    """
    Crontab expression month part.
    """


@dataclasses.dataclass(frozen=True)
class CronPartWeekday(
    CronPart,
    name="Weekday",
    min_value=1,
    max_value=7,
    aliases={
        "MON": 1,
        "TUE": 2,
        "WED": 3,
        "THU": 4,
        "FRI": 5,
        "SAT": 6,
        "SUN": 7,
        # Spec is 0 or 7 support is Sunday.
        # This ensures that we only output one option if both provided.
        "0": 7,
    },
):
    """
    Crontab expression weekday part.
    """


@dataclasses.dataclass(frozen=True)
class Crontab:
    """
    Crontab Expression class to store the results of a parsed
    cron expression (incl command).

    Cron expression part values are expanded to be complete list. E.G.
    */15 = 0 15 30 45
    """

    minute: CronPartMinute
    hour: CronPartHour
    monthday: CronPartMonthday
    month: CronPartMonth
    weekday: CronPartWeekday
    tz: dt.tzinfo
    command: str

    def __str__(self) -> str:
        return " ".join(
            [
                f"minute={self.minute}",
                f"hour={self.hour}",
                f"monthday={self.monthday}",
                f"month={self.month}",
                f"weekday={self.weekday}",
                f"tz={self.tz}",
                f"command={self.command}",
            ]
        )

    @classmethod
    def from_parse(
        cls,
        *,
        expr: str,
        tz: dt.tzinfo,
        now: dt.datetime | None = None,
    ) -> Crontab:
        """
        Parses Crontab expression which also includes the command to run.
        """
        # Resolve macros (@weekly, @daily etc) to equivalent cron expressions.
        # Split the expression to see if it contains a macro in the first indices.
        # This would be the case if a macro and command was passsed in like "@annually /usr/bin/find"
        x = expr.split()
        try:
            macro_val = CRON_MACROS[x[0]]
            value = f"{macro_val} {x[1]}"
        except KeyError:
            value = expr

        # 5 for cron schedule + 1 for cron command = 6
        fields_len_constraint = 6
        fields = value.split()
        fields_len = len(fields)
        if fields_len != fields_len_constraint:
            raise ValueError(
                f"{cls.__qualname__} expression must be of {fields_len_constraint} fields, "
                f"Received: {fields_len}"
            )

        fields_iter = iter(fields)
        now = (now or dt.datetime.now(tz=tz)).astimezone(tz)

        return cls(
            minute=CronPartMinute.from_expr(next(fields_iter)),
            hour=CronPartHour.from_expr(next(fields_iter)),
            monthday=CronPartMonthday.from_expr(next(fields_iter)),
            month=CronPartMonth.from_expr(next(fields_iter)),
            weekday=CronPartWeekday.from_expr(next(fields_iter)),
            command=str(next(fields_iter)),
            tz=now.tzinfo or tz,
        )

    @property
    def next_scheduled_run(self) -> dt.datetime:
        return next(self.iter())

    def iter(self, start: dt.datetime | None = None) -> Iterator[dt.datetime]:
        """
        Yields future schedules for this crontab expression.
        """
        anchor = start.astimezone(self.tz) if start else dt.datetime.now(tz=self.tz)
        anchor_date = anchor.date()

        for day_date in self._generate_future_dates(anchor_date):
            is_start_day = day_date == anchor_date

            for valid_hour in self.hour:
                is_start_hour = is_start_day and valid_hour == anchor.hour

                # If today is the date the job should start but the hour has
                # passed we need to skip
                if is_start_day and valid_hour < anchor.hour:
                    continue

                for valid_minute in self.minute:
                    # If today is the date the job and the hour it should start
                    # but the minute has passed we need to skip.
                    if is_start_hour and valid_minute < anchor.minute:
                        continue

                    yield dt.datetime(
                        year=day_date.year,
                        month=day_date.month,
                        day=day_date.day,
                        hour=valid_hour,
                        minute=valid_minute,
                        tzinfo=self.tz,
                    )

    def _generate_future_dates(self, start: dt.date | None = None) -> Iterator[dt.date]:
        """
        Yields future dates for the crontab expression based on the
        month, monthday and weekdays parts.
        """
        cal = calendar.Calendar()
        anchor = start if start else dt.date.today()

        # Cap generation at year 2099, this will give us good buffer.
        while anchor.year < 2099:
            for valid_month in self.month:
                # Check if the month is lower the current month.
                # If so we need to skip because it will be in the past.
                if valid_month < anchor.month:
                    continue

                for day_no, weekday_no in cal.itermonthdays2(anchor.year, valid_month):
                    # itermonthdays yields all days required to get a full week,
                    # but those outside the current month are 0 so we should ignore
                    # those.
                    if day_no == 0:
                        continue

                    # It is the same month, but the day is in the past so we should skip.
                    if valid_month == anchor.month and day_no < anchor.day:
                        continue

                    # Check if this weekday is valid.
                    # In the calendar module, dates are 0-based. Monday == 0 and Sunday == 6.
                    # Our approach is 1-based so we need to add 1 to the value to lookup correctly
                    if weekday_no + 1 not in self.weekday:
                        continue

                    # Check if this monthday is valid.
                    if day_no not in self.monthday:
                        continue

                    # Success! Valid date.
                    yield dt.date(year=anchor.year, month=valid_month, day=day_no)

            # After exhausting each month for the current anchor year, we need
            # to start again from a new year.
            anchor = dt.date(year=anchor.year + 1, month=1, day=1)
